### PR TITLE
feat: get wallet api

### PR DESCRIPTION
**Pull request by alaibe**
Build a new endpoint for the wallet which include all necessary data to display the wallet

### CODE REVIEW

File: services/wallet/reader.go [hunk 1]
```diff
@@ -0,0 +1,238 @@
...
+
+type ReaderAccount struct {
+	Account      *accounts.Account               `json:"account"`
+	Collections  map[uint64][]*OpenseaCollection `json:"collections"`
+	Tokens       map[uint64][]*ReaderToken       `json:"tokens"`
+	Transactions map[uint64][]*transfer.View     `json:"transactions"`
+}
+
+type Wallet struct {
+	Accounts            []*ReaderAccount                 `json:"Accounts"`
```
Comment (richard-ramos) [inline services/wallet/reader.go]:
```suggestion
	Accounts            []*ReaderAccount                 `json:"accounts"`
```

Ignore this suggestion if this is already used by mobile

File: services/wallet/reader.go [hunk 2]
```diff
@@ -0,0 +1,238 @@
...
+	for _, chainID := range chainIDs {
+		client, err := newOpenseaClient(chainID, r.s.openseaAPIKey)
+		if err == nil {
+			c, _ := client.fetchAllCollectionsByOwner(common.Address(account.Address))
+			collections[chainID] = c
+		}
+
+		for _, token := range visibleTokens {
+			oraclePrice := prices[token.Symbol]
+			cryptoBalance := balances[token.Address].ToInt()
```
Comment (richard-ramos) [inline services/wallet/reader.go:83]: Could it be possible that the balance for a token does not exist in the map, and we run into a null exception when calling `.ToInt()`?

Comment (alaibe) [reply services/wallet/reader.go:83]: That shouldn't happen, at least I don t see how, as we are fetching the balances for all addresses/all visible tokens. So should be at least 0

**Review by Khushboo-dev-cpp — APPROVED** [review]
Looks really really good to me! excited to see how the calls are super simplified on the BE side :)

File: services/wallet/api.go [hunk 3]
```diff
@@ -223,11 +239,11 @@ func (api *API) AddFavourite(ctx context.Context, favourite Favourite) error {
 	return err
 }
 
-func (api *API) GetCryptoOnRamps(ctx context.Context) ([]CryptoOnRamp, error) {
+func (api *API) GetCryptoOnRamps(ctx context.Context) ([]*CryptoOnRamp, error) {
```
Comment (Samyoul) [inline services/wallet/api.go]: I'm curious, what benefit do we get from passing a slice of references over a slice of values?

File: services/wallet/api.go [hunk 4]
```diff
@@ -6,18 +6,34 @@ import (
...
 	"github.com/ethereum/go-ethereum/log"
+	"github.com/status-im/status-go/multiaccounts/accounts"
 	"github.com/status-im/status-go/params"
 	"github.com/status-im/status-go/services/wallet/chain"
 	"github.com/status-im/status-go/services/wallet/transfer"
 )
 
 func NewAPI(s *Service) *API {
-	return &API{s}
+	accountDb, _ := accounts.NewDB(s.db)
```
Comment (Samyoul) [inline services/wallet/api.go]: Why are we ignoring the error here?

File: services/wallet/service.go [hunk 5]
```diff
@@ -38,6 +39,7 @@ func NewService(db *sql.DB, rpcClient *rpc.Client, accountFeed *event.Feed, open
 
 // Service is a wallet service.
 type Service struct {
+	db                    *sql.DB
```
Comment (Samyoul) [inline services/wallet/service.go:44]: Service accepts a `*sql.DB`, but then only uses it to generate an `accounts.Database`, perhaps `Service` should accept just the `accounts.Database`

**Review by Samyoul — CHANGES_REQUESTED** [review]

File: services/wallet/api.go (hunk 4 above)
Comment (alaibe) [reply services/wallet/api.go]: Updating by using the accountsdb given by the service

File: services/wallet/service.go (hunk 5 above)
Comment (alaibe) [reply services/wallet/service.go:44]: It doesn t only use for generating an accounts.DB but I pass specifically the accounts db being build at the layer above

File: services/wallet/api.go (hunk 3 above)
Comment (alaibe) [reply services/wallet/api.go]: Isn t it better to pass the pointer rather than copy the whole CryptoOnRamp?

Comment (Samyoul) [reply services/wallet/api.go]: No, a slice index is a pointer effectively. A slice pointer value is much slower.

**Review by richard-ramos — APPROVED** [review]

File: services/wallet/api.go (hunk 3 above)
Comment (alaibe) [reply services/wallet/api.go]: TIL, I will update the PR accordingly, thanks

**Comment by alaibe** [comment]
@Samyoul PR has been updated

**Review by Samyoul — APPROVED** [review]
All looks great, thank you. Also have a little look at this article https://medium.com/@philpearl/bad-go-slices-of-pointers-ed3c06b8bb41 . It has got some benchmark examples of how a slice of pointer values impacts performance.

**Pull request merged**